aptos-move/framework/aptos-framework/sources/account.move

```diff
@@ -249,6 +254,77 @@ module aptos_framework::account {
+            auth_key = multi_ed25519::unvalidated_public_key_to_authentication_key(&pk);
+        };
+        auth_key
+    }
+
+    public entry fun rotate_authentication_key(
+        account: &signer,
+        from_scheme: u8,
+        from_public_key_bytes: vector<u8>,
+        to_scheme: u8,
+        to_public_key_bytes: vector<u8>,
+        cap_rotate_key: vector<u8>,
+        cap_update_table: vector<u8>,
+    ) acquires Account, OriginatingAddress {
+        let addr = signer::address_of(account);
+        assert!(exists_at(addr), error::not_found(EACCOUNT_DOES_NOT_EXIST));
+        assert!((from_scheme == ED25519_SCHEME || from_scheme == MULTI_ED25519_SCHEME), EINVALID_SCHEME);
+        assert!((to_scheme == ED25519_SCHEME || to_scheme == MULTI_ED25519_SCHEME), EINVALID_SCHEME);
+
+        let account_resource = borrow_global_mut<Account>(addr);
+
+        // verify the public key matches the current authentication key
+        if (from_scheme == ED25519_SCHEME) {
+            let from_pk = ed25519::new_unvalidated_public_key_from_bytes(from_public_key_bytes);
+            let from_auth_key = ed25519::unvalidated_public_key_to_authentication_key(&from_pk);
+            assert!(account_resource.authentication_key == from_auth_key, error::unauthenticated(EWRONG_CURRENT_PUBLIC_KEY));
+        } else {
+            let from_pk = multi_ed25519::new_unvalidated_public_key_from_bytes(from_public_key_bytes);
+            let from_auth_key = multi_ed25519::unvalidated_public_key_to_authentication_key(&from_pk);
+            assert!(account_resource.authentication_key == from_auth_key, error::unauthenticated(EWRONG_CURRENT_PUBLIC_KEY));
+        };
+
+        let curr_auth_key = from_bcs::to_address(account_resource.authentication_key);
+        let challenge = RotationProofChallenge {
+            sequence_number: account_resource.sequence_number,
+            originator: addr,
+            current_auth_key: curr_auth_key,
+            new_public_key: to_public_key_bytes,
+        };
+
+        // verify the signatures from the current private key and the new private key are valid
+        let curr_auth_key = verify_key_rotation_signature(from_scheme, from_public_key_bytes, cap_rotate_key, &challenge);
+        let new_auth_key = verify_key_rotation_signature(to_scheme, to_public_key_bytes, cap_update_table, &challenge);
+
+        // update the address_map table, so that we can reference to the originating address using the current address
+        let address_map = &mut borrow_global_mut<OriginatingAddress>(@aptos_framework).address_map;
+        let originating_address = addr;
+        let curr_address = from_bcs::to_address(curr_auth_key);
+        let new_address = from_bcs::to_address(new_auth_key);
+        if (table::contains(address_map, curr_address)) {
+            originating_address = table::remove(address_map, curr_address);
```

Comment:
I'm a bit confused by this: shouldn't this be an `assert!(originating_address == table::remove(address_map, curr_address), 1);`? 

In other words, this assigns the same value to `originating_address`, no?

In other _other_ words, whatever originating address we remove from table[curr_auth_key], should be the original address of the signer computed on line 282 at the beginning of `rotate_authentication_key`, no? After all, the signer's account has `curr_auth_key` set as its current authentication key (as per the check on line 293).

Maybe I'm missing something / too tired.

Reply:
ah yea I think you're right. 

I was thinking of the usecase where a user rotates from `key_a` to `key_b`, and now they want to rotate to `key_c`. if they call the function from `account_b` - it's not the originating account so the address of that account is not the `originating_address`, and in that case the `originating_address` would be the value we removed from the table? but I think in general, we're expecting the user to call this function from `account_a`. 

just updated this part of code.

Reply:
why do we need to enforce user have to rotate from account_a? previous code works when user rotate from both account_a and account_b right.

Reply:
I guess we want some kind of standard? and it is more intuitive to rotate it from the original account

Reply:
Uhm.... Now, I'm confused: when you rotate to `key_b` you don't create an `account_b` at all. You just update `account_a`'s `authentication_key`. And you update the table, to make `key_b` point to `account_a`'s address.

So what do you mean by "if they call the function from `account_b`", if there's no `account_b`.

@davidiw can you help us here?

Also, @davidiw, what happens if somebody tries to create this non-existent `account_b` **after** the rotation to `key_b` happened?

I think that might cause the problem @chloeqjz is talking about above? Or it might cause other problems...

Reply:
@alinush `I thought when you rotate to key_b, you don't create an account_b at all.` this is no enforcement that the person doesn't have account_b right. In fact if they own the private key to account b, they own account_b in fact.  I think the assumption that they won't or shouldn't call the function from account_b is not valid assumption. make sense?

Reply:
After clarifying in person/over Slack, here’s the problematic scenario that this PR fixes via [the assert here](https://github.com/aptos-labs/aptos-core/pull/3626/files#diff-ef8a70cd4b03dc7b8b6dfd8306886d480c155211d8c472b59585471e4c60c427R283) 🧯:
1. User rotated `acc_a`’s `key_a` to `key_b`.
2. User creates an account `acc_b` with `key_b` as its authentication key (which is mapped to `acc_a` via the originating address table)
3. User calls rotate_key from `acc_b` (rather than `acc_a`) asking to rotate `acc_b`’s `key_b` to `key_c`

This is where there is a bug. Specifically, our old `rotate_authentication_key_ed25519` implementation (as well as our initial implementation of `rotate_authentication_key` in the current PR) does the following:
1. Fetches `addr_a = OriginatingAddress[key_b]` (See [this change](https://github.com/aptos-labs/aptos-core/pull/3483/files#diff-ef8a70cd4b03dc7b8b6dfd8306886d480c155211d8c472b59585471e4c60c427R238).)
2. Removes `OriginatingAddress[key_b]` (See [the same change](https://github.com/aptos-labs/aptos-core/pull/3483/files#diff-ef8a70cd4b03dc7b8b6dfd8306886d480c155211d8c472b59585471e4c60c427R238).)
3. Sets `OriginatingAddress[key_c] = addr_a`
4. Sets `acc_b`’s (not `acc_a`’s) authentication key to `key_c`. (See [this line.](https://github.com/aptos-labs/aptos-core/pull/3483/files#diff-ef8a70cd4b03dc7b8b6dfd8306886d480c155211d8c472b59585471e4c60c427R249))

As a result, we’ve left the system in an inconsistent state:
1. `acc_a` still has `key_b` as its auth key, but no originating-address mapping
2. `acc_b` has `key_c` as its auth key, with an originating-address mapping it back to `addr_a`.

In other words, the originating address table and the resource at `acc_a` don’t agree on what the current auth key for `acc_a` is.

One way to deal with this is to prevent key rotation from `acc_b` via the assert Chloe just introduced [here](https://github.com/aptos-labs/aptos-core/pull/3626/files#diff-ef8a70cd4b03dc7b8b6dfd8306886d480c155211d8c472b59585471e4c60c427R325). (edited)